compat/mingw.c

```diff
@@ -475,6 +475,50 @@ static wchar_t *normalize_ntpath(wchar_t *wbuf)
 	return wbuf;
 }
 
+/* rename and unlink a file. This is needed for the ability to recreate a file
+ * with the same name, for example on checkout which unlinks and recreates the file */
+static int rename_and_unlink(const wchar_t *pathname, int (*unlinker)(const wchar_t *))
+{
+	int res;
+	/* leave some space for the .unlink_XXXXXX extension */
+	wchar_t temppath[MAX_LONG_PATH + 15];
+	*temppath = 0;
+
+	wcsncat(temppath, pathname, sizeof(temppath));
+	wcsncat(temppath, L".unlink_XXXXXX", sizeof(temppath));
+	if (!_wmktemp(temppath) || !MoveFileExW(pathname, temppath, 0))
```

Comment:
With paths close to `MAX_PATH`, this `temppath` might run over... so I thought about using the Recycle Bin instead, but that would probably complicate the code too much.

But still, I wonder whether we have to see whether `temppath` is now greater than, or equal to, `MAX_PATH` and do the `\\?\` dance in that case?

Also: I am slightly concerned that the performance might suffer from first calling `CreateFile()` and then `MoveFile()` and then `DeleteFile()`. Creating a handle to a file is not cheap. Certainly not when you do it for 90k files.

So I wonder whether you have thought about using [`SetFileInformationByHandle()`](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-setfileinformationbyhandle) to rename and to delete the file here?

Reply:
I reach here only if I did *not* receive a HANDLE because the file failed to open.

So I'll need to call `CreateFile` anyway for deleting it, which is quite similar to what `_wunlink` does.

We could however replace the dual `unlinker` calls by using `CreateFile` with:
```
FILE_ATTRIBUTE_NORMAL
| FILE_ATTRIBUTE_DIRECTORY
| FILE_FLAG_BACKUP_SEMANTICS
| FILE_FLAG_OPEN_REPARSE_POINT
| FILE_FLAG_DELETE_ON_CLOSE
```
followed by `CloseHandle`. But that's for a separate patch...

Reply:
By the way, notice that MoveFile/DeleteFile only happens for files that are open by other processes. This should be typically up to 10s of files, certainly not 10Ks.

Reply:
Yes, you're right. My only remaining concern is the `MAXPATH` one...

Reply:
The previous revisions were limited to `MAX_PATH-15` from the original file name (which could land in the parent directory, but this is typically on the same filesystem, so it shouldn't matter...) Do you prefer that, or using `\\?\`?

Reply:
Yes, I think I do prefer the `-15` version. Sorry for missing that.

Reply:
Done + Rebased.

Reply:
Excellent, thanks!